File: src/algorithms/draft/leakyBucket.ts
// leaky bucket as meter, leaky bucket as queue

class LeakyBucket {
	public capacity: number;
	public drops: number;
	public lastUpdatedAt: number;
	public leakRate: number;

	public consume(key: string, cost: number) {
		const now = Date.now();
		const millisecondsElapsed = now - this.lastUpdatedAt;
		this.lastUpdatedAt = now;

		// Decrement drops leaked since last call
		// todo: 0 guard
		this.drops -= this.leakRate * millisecondsElapsed;

		const nextTokens = this.drops + cost;
		if (nextTokens <= this.capacity) {
			this.drops = nextTokens;
			return true;
		}

		return false;
	}
}

File: src/algorithms/draft/leakybucket.lua
-- Redis script to implement a leaky bucket
-- see https://medium.com/callr-techblog/rate-limiting-for-distributed-systems-with-redis-and-lua-eeea745cb260

local capacity    = tonumber(ARGV[1]) -- max number of tokens bucket can contain
local leakRate    = tonumber(ARGV[2]) -- number of tokens to drain each interval
local cost        = tonumber(ARGV[3]) -- how many tokens to consume, default is 1
local now         = tonumber(ARGV[4]) -- current unix time in seconds

-- get bucket?

-- elapsed calculated as now - lastUpdated

-- decrement count by leak rate
-- this.drops -= this.leakRate * millisecondsElapsed;

-- calculate next tokens by cost

-- if next > capacity, overflow

-- allow


-- (c) Florent CHAUVEAU <florent.chauveau@gmail.com>

local now  = tonumber(ARGV[1])
local cps = tonumber(ARGV[2])
local key = KEYS[1]

-- remove tokens < min (older than now() -1s)
redis.call('ZREMRANGEBYSCORE', key, '-inf', now - 1)

local last = redis.call('ZRANGE', key, -1, -1)

local next = now

-- if there are entries to count
if type(last) == 'table' and #last > 0 then
  for key,value in pairs(last) do
    next = tonumber(value) + 1 / cps
    break -- break at first item
  end
end

if now > next then
  -- the current timestamp is > than last+1/cps
  -- we'll keep now
  next = now
end

redis.call('ZADD', key, next, next)

return tostring(next - ts)
